pallets/aura-ext/src/consensus_hook.rs

```diff
@@ -77,3 +80,46 @@ where
 		)
 	}
 }
+
+impl<
+		T: pallet::Config + parachain_system::Config,
+		const RELAY_CHAIN_SLOT_DURATION_MILLIS: u32,
+		const V: u32,
+		const C: u32,
+	> FixedVelocityConsensusHook<T, RELAY_CHAIN_SLOT_DURATION_MILLIS, V, C>
+{
+	/// Whether it is legal to extend the chain, assuming the given block is the most
+	/// recently included one as-of the relay parent that will be built against, and
+	/// the given slot.
+	///
+	/// This should be consistent with the logic the runtime uses when validating blocks to
+	/// avoid issues.
+	///
+	/// When the unincluded segment is empty, i.e. `included_hash == at`, where at is the block
+	/// whose state we are querying against, this must always return `true` as long as the slot
+	/// is more recent than the included block itself.
+	pub fn can_build_upon(included_hash: T::Hash, new_slot: Slot) -> bool {
+		let velocity = V.max(1);
+		let (last_slot, authored_so_far) = match pallet::Pallet::<T>::slot_info() {
```

Comment:
Capacity-related code actually panics if this value isn't present in the storage, because it means that the aura-ext pallet was never initialized.

There's also a condition below when slot moved backwards, which also shouldn't be possible.

To me it looks like we simply avoid panicking in this function, perhaps because it's going to be part of runtime API. In the end, this is super edge-case which makes little to no difference. As a safeguard `false` could make more sense.

Reply:
I believe everyone should start with a simple `FixedCapacity` hook to stay backwards compatible.

Reply:
I agree. Good coding practice is to minimize the assumptions you make in a piece of code, especially w.r.t. how it might be used. I just follow this rule of thumb whenever I'm writing code and it leads to fewer bugs in a large codebase.